lib/phoenix_ecto/plugs/check_repo_ready.ex

```diff
@@ -0,0 +1,64 @@
+defmodule Phoenix.Ecto.Plugs.CheckRepoReady do
+  @moduledoc """
+  A plug that does some checks on your application repos.
+
+  Checks if the storage is up (database is created) or if there are any pending migrations.
+  Both checks can raise an error if the conditions are not met.
+
+  ## Plug options
+
+  * `:otp_app` - name of the application which the repos are fetched from
+  * `:get_migration_function` - function that returns the migrations for a given `repo`.
+    Should return a list of tuples with three elements,
+    e.g. `[{state :: :up | :down, version :: integer(), name :: String.t()]]`
+    If `ecto_sql` dependecy is loaded, uses `Ecto.Migrator.migrations/1` by default.
+  """
+
+  @behaviour Plug
+
+  alias Plug.Conn
+
+  def init(opts), do: opts
+
+  def call(%Conn{} = conn, opts) do
+    repos = Application.get_env(opts[:otp_app], :ecto_repos, [])
+
+    for repo <- repos, Process.whereis(repo) do
+      check_storage_up!(repo)
+      check_pending_migrations!(repo, opts)
+    end
+
+    conn
+  end
+
+  defp check_storage_up!(repo) do
+    adapter_implement_storage_status? =
+      Code.ensure_loaded?(repo.__adapter__) &&
+        function_exported?(repo.__adapter__, :storage_status, 1)
+
+    storage_status_down? = fn -> repo.__adapter__.storage_status(repo.config()) == :down end
+
+    if adapter_implement_storage_status? && storage_status_down?.(),
+      do: raise(Phoenix.Ecto.StorageNotCreatedError, repo: repo)
+  end
+
+  defp check_pending_migrations!(repo, opts) do
+    ecto_sql_loaded? = Code.ensure_loaded?(Ecto.Migrator)
+
+    # If the dependency `ecto_sql` is not loaded we can't check if
+    # there are pending migrations so we try to fail gracefully here
+    fallback_get_migrations_function =
```

Comment:
I thought about doing something like this so we don't rerun this check every time the plug is called, would it be a nice idea? 🤔

```elixir
defmodule Phoenix.Ecto.Plugs.CheckRepoReady do
  # ...
  @fallback_get_migrations_function if Code.ensure_loaded?(Ecto.Migrator),
                                      do: &Ecto.Migrator.migrations/1,
                                      else: &__MODULE__.noop_get_migrations_function/1
  # ...
  defp check_pending_migrations!(repo, opts) do
    get_migrations_function =
      Keyword.get(opts, :get_migrations_function, @fallback_get_migrations_function)
  # ...
  end

  defp noop_get_migrations_function(_repo), do: []
end

Reply:
It is a cheap check, so I wouldn't worry.

Reply:
Cool cool cool 😄